pkg/apis/core/v1alpha1/tiltrun_types.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2020 The Tilt Dev Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/tilt-dev/tilt-apiserver/pkg/server/builder/resource"
+	"github.com/tilt-dev/tilt-apiserver/pkg/server/builder/resource/resourcestrategy"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TiltRun provides introspective data about the status of the Tilt process.
+// +k8s:openapi-gen=true
+type TiltRun struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TiltRunSpec   `json:"spec,omitempty"`
+	Status TiltRunStatus `json:"status,omitempty"`
+}
+
+// TiltRunList is a list of TiltRun objects.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type TiltRunList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []TiltRun `json:"items"`
+}
+
+// TiltRunSpec defines the desired state of TiltRun
+type TiltRunSpec struct {
+	// TiltfilePath is the path to the Tiltfile for the run. It cannot be empty.
+	TiltfilePath string `json:"tiltfilePath"`
+	// ExitCondition defines the criteria for Tilt to exit.
+	ExitCondition ExitCondition `json:"exitCondition"`
+}
+
+type ExitCondition string
+
+const (
+	// ExitConditionManual cedes control to the user and will not exit based on resource status.
+	// This is used by `tilt up`.
+	ExitConditionManual ExitCondition = "manual"
+	// ExitConditionFirstFailure terminates upon the first encountered build or runtime failure.
```

Comment:
i'm not sure I would call this FirstFailure, since it will also exit if all servers are ready and all jobs are complete?

Reply:
That's a great point 🤔 

One possible change would be just make this a mouthful but more descriptive (`first-failure-or-all-succeeded`).

I briefly thought about splitting this into `FailureCondition` + `SuccessCondition` but you can come up with weird combinations (e.g. success of "all succeeded" but failure of "never" is probably _not_ what you want).

I'm now wondering if this could be something like `[]ExitCondition ExitConditions` (where the list is OR'd, i.e. first to hit) and then have a corresponding status field of `TriggeredExitCondition`. For `tilt ci`, you'd have `spec.exitConditions: ["first-failure", "all-successful"]`.

That said, it still introduces awkwardness with `manual` (or maybe more accurately that should be called `user-signal`) - it's _always_ a valid exit condition, so in the `tilt up` case, should the exit condition list just be nil/empty?

Reply:
i honestly think ExitConditionCI, since it has CI-like semantics, and "CI" is the term of art people are going to be searching for when they want this